_posts/2022-05-31-introduction-to-project-panama-part-1.markdown

```diff
@@ -121,86 +119,98 @@ FunctionDescriptor printfDescriptor = FunctionDescriptor.of(
 We need somehow to bind Java objects to native memory segments to make sure C _printf_ can access them.
-Here’s the moment when we step into the area of **Foreign Memory Access API** (FMA)!
 
-Way back in time when I did C programming (in high school, honestly), a memory allocation and freeing were kinda painful
-because no matter how good you are you’d, eventually you will forget to allocate or free a memory which will cause a
+Both memory allocation and freeing were painful because no matter how good software developer you are, eventually you will forget to allocate or free a memory which will cause a
 program to leak it or blow up with a segmentation fault.
 
-In Java, we traditionally relied on the GC to take care of memory cleaning up.
+Java relies on a Garbage Collector to allocate and free memory but Panama’s FMA API allocates memory off-heap.
 In Panama, the FMA API helps to allocate a memory off-heap which is a crucial part of any native interop story!
 
 The FMA API allows to allocate and access memory segments, their addresses, and the shape of contiguous memory
 regions, located either on- or off- the Java heap.
-In concepts of Panama, all allocated memory segments must be bound to a specific memory session (**MemorySession**),
-in other terms, consider a session like a unified memory allocation tool,
-for instance, C `malloc`, but auto-closable (thanks to Java try-with-resources syntax block).
+All allocated memory segments are bound to a specific memory session (**MemorySession**), 
+an instance of it offers a set of APIs to allocate memory segments. Consider a memory session like a unified memory allocation tool,
+like C `malloc`.
+One of the important features of **MemorySession** is an implementation of an **AutoClosable** interface - 
+it is better to use it within the try-with-resources syntax code block.
 
-Last, but not least important part of a program is a segment allocator derived from a memory session.
-By itself, a memory session serves the role of a native memory allocation API:
+Last, but not least important part of any program performing downcalls is an instance of **SegmentAllocator**:
 ```java
 try (var memorySession = MemorySession.openConfined()) {
     // memory allocation happening here
     SegmentAllocator allocator = SegmentAllocator.newNativeArena(memorySession);
 }
 ```
 
-As I was told by Panama developers, there are so many ways to do what you need in the right way, especially in terms of memory allocation.
-One of the possible memory allocation methods in Java is **SegmentAllocator**, which is similar to **MemorySession**,
+There are more than one right way to allocate a memory segment, 
+one of the possible memory allocation methods in Java is **SegmentAllocator**, which is similar to **MemorySession**,
 but often used in the context of other C memory allocation functions different from C [malloc](https://en.cppreference.com/w/c/memory/malloc).
```

Comment:
I can't parse _but often used in the context of other C memory allocation functions different from C [malloc]_